physical/etcd.go

```diff
@@ -167,3 +179,188 @@ func (b *EtcdBackend) nodePath(key string) string {
+	// resulting key as a "semaphore key".
+	// https://coreos.com/etcd/docs/2.0.8/api.html#atomically-creating-in-order-keys
+	response, err := c.backend.client.CreateInOrder(c.backend.nodePathLock(c.key), "Vault Lock", EtcdLockTTL)
+	if err != nil {
+		return "", 0, err
+	}
+
+	return response.Node.Key, response.EtcdIndex, nil
+}
+
+// getSemaphoreKey determines which semaphore key holder has aquired the lock.
+func (c *EtcdLock) getSemaphoreKey() (string, uint64, error) {
+
+	// Get the list of waiters in order to see if we are next.
+	response, err := c.backend.client.Get(c.backend.nodePathLock(c.key), true, false)
+	if err != nil {
+		return "", 0, err
+	}
+
+	// Make sure the list isn't empty.
+	if response.Node.Nodes.Len() == 0 {
+		return "", response.EtcdIndex, nil
+	}
+
+	return response.Node.Nodes[0].Key, response.EtcdIndex, nil
+}
+
+func (c *EtcdLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+
+	// Get the local lock before interacting with etcd.
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// Check if the lock is already held.
+	if c.isHeld {
+		return nil, EtcdLockHeldError
+	}
+
+	// Add a new semaphore key that we will track.
+	semaphoreKey, _, err := c.addSemaphoreKey()
+	if err != nil {
+		return nil, err
+	}
+	c.semaphoreKey = semaphoreKey
+
+	// Get the current semaphore key.
+	currentSemaphoreKey, currentEtcdIndex, err := c.getSemaphoreKey()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create an etcd-compatible boolean stop channel from the provided
+	// interface stop channel.
+	boolStopCh := make(chan bool)
+	go func() {
+		for _ = range stopCh {
+		}
+		close(boolStopCh)
+	}()
+
+	// Loop until the we current semaphore key matches ours.
+	for semaphoreKey != currentSemaphoreKey {
+		var err error
+
+		// Start a watch of the entire lock directory, providing the stop channel.
+		response, err := c.backend.client.Watch(c.backend.nodePathLock(c.key), currentEtcdIndex+1, true, nil, boolStopCh)
+		if err != nil {
+
+			// If the error is not an etcd error, we can assume it's a notification
+			// of the stop channel having closed. In this scenario, we also want to
+			// remove our semaphore key as we are no longer waiting to aquire the
+			// lock.
+			if _, ok := err.(*etcd.EtcdError); !ok {
+				_, err = c.backend.client.Delete(c.semaphoreKey, false)
+			}
+
+			return nil, err
+		}
+
+		// Make sure the index we are waiting for has not been removed. If it has,
+		// this is an error and nothing else needs to be done.
+		if response.Node.Key == semaphoreKey &&
+			(response.Action == "delete" || response.Action == "expire") {
+			return nil, EtcdSemaphoreKeyRemovedError
+		}
+
+		// Get the current semaphore key and etcd index.
+		currentSemaphoreKey, currentEtcdIndex, err = c.getSemaphoreKey()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// We now hold the lock.
+	c.isHeld = true
+
+	// Because we hold the lock, we can set our value for the given key.
+	if err := c.backend.Put(&Entry{
```

Comment:
The race will occur between the other standby instances. E.g. they will see the lock has happened, but not have the associated value which will cause errors on their end.